NumericWidget : Mouse wheel adjust value

ericmehl opened the pull request:
This adds the ability to change a numeric widget value with the mouse wheel in the same manner as the keyboard up and down keys currently do.

I arrived at the scale factor of `15.0` to translate the `wheelRotation` value to increment units by seeing what the `wheelRotation` value was for a single "tick" on my mouse. I don't know how universal that is, so could definitely do with some testing by others to see if it changes values in a reasonable way.

Fixes #6009.

### Checklist ###

- [X] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/main/CONTRIBUTING.md).
- [X] I have updated the documentation, if applicable.
- [X] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [X] My code follows the Gaffer project's prevailing coding style and conventions.

murraystevenson commented on Changes.md:
```diff
@@ -19,6 +19,7 @@ Improvements
   - Improved layout of Box node plug creator visibility toggles.
 - ArnoldShader : Moved the `toon` shader's `*_tonemap_hue_saturation` parameters to appropriate sections in the UI.
 - File Browser : The "Type" column can now be sorted. This sorts directories separately from files, which are sorted by their extension.
+- NumericWidget : Added the ability to use the mouse wheel to adjust values in the same manner as <kbd>Up</kbd> and <kbd>Down</kbd> (#6009).
```
This will need to move up to 1.4.x.x as 1.4.12.0 has been released.

ericmehl replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/6045/commits/c51f7e045ba402ab805ff739c82c3b3577f53c10

murraystevenson reviewed (changes requested):
Thanks Eric! The scale factor of 15 feels about right for my Logitech mouse on Linux, I see roughly 1 (+/- 0.125) increment's worth of change for each "step" of the wheel. I've added a couple of thoughts inline based on some early testing.

johnhaddon commented on python/GafferUI/NumericWidget.py:
```diff
@@ -141,6 +142,19 @@ def __keyPress( self, widget, event ) :
+
+		scaleFactor = 15.0
+
+		self.__incrementIndex( self.getCursorPosition(), event.wheelRotation / scaleFactor )
```
I have a suggestion which fixes this but also removes any concerns about different mouse-wheel speeds :

`self.__incrementIndex( self.getCursorPosition(), int( math.copysign( 1, event.wheelRotation ) ) )`

murraystevenson replied:
It might just be my setup but the downside to "any scroll amount is considered one" is that I see 3-5 wheel events for any one "step" of my scroll wheel so I'm unable to make fine adjustments. I'm happy to blame it on my mouse if this is working acceptably for you both...

ericmehl commented on python/GafferUI/NumericWidget.py:
```diff
@@ -141,6 +142,19 @@ def __keyPress( self, widget, event ) :
+		assert( widget is self )
+
+		if not self.getEditable() :
+			return False
```
That's a good idea, I added the check for focus and that `Ctrl` is pressed in https://github.com/GafferHQ/gaffer/pull/6045/commits/bf5886754d71dc70df8d281fd405575d8a0ffc84

murraystevenson replied:
Thanks, the interaction does feel a lot more intentional with the focus and `Ctrl` checks in place.

One thing we'll still need to handle is the mouse wheel event falling through and scrolling the parent while you're attempting to adjust a value. This should be fairly apparent if you try using the scroll wheel to adjust a parameter on a shader like `standard_surface` or `principled_bsdf` while a few of the sections are expanded and the Node Editor's vertical scroll bar is visible. 

A solution for this might just be to stick a `qEvent.accept()` in: https://github.com/GafferHQ/gaffer/blob/cbd8e39a663cb9a17b3fdb15c59c2e848ba73d9a/python/GafferUI/Widget.py#L1271-L1281

ericmehl replied:
Thanks for pointing that out in `Widget`, I would have struggled a bit to locate where the signal was falling through. I don't understand why we need this for the wheel event but not others (except the foreign drag), but it works. And it doesn't break the other uses of `wheelSignal()` like viewport zooming.

Done in https://github.com/GafferHQ/gaffer/pull/6045/commits/4203049af5a41a9948e961aaccca3995f747d697. I assume based on other uses of `accept()` that we only want to accept if the signal handler returns `True`.

murraystevenson replied:
Thanks, that's done the trick.

ericmehl commented:
Thanks @murraystevenson and @johnhaddon! I addressed the comments and also tested on a laptop track pad. It may be a touch too fast using a track pad, but I was still able to increment one at a time without much trouble. Since the mouse case feels natural, I'm inclined to leave it at the current rate.

ericmehl commented:
The latest push refines the increment logic and also fixes the scroll signal leaking. In https://github.com/GafferHQ/gaffer/pull/6045/commits/cb7e9349176676e0524854bfa8c15d56eda22949 I added to the tooltip to try and make this a little more discoverable.

Ready for more testing and review!

murraystevenson commented on python/GafferUI/NumericPlugValueWidget.py:
```diff
@@ -77,7 +77,7 @@ def getToolTip( self ) :
 				result += "\n"
 			result += "## Actions\n"
-			result += " - Cursor up/down to increment/decrement\n"
+			result += " - Cursor or scroll wheel up/down to increment/decrement\n"
```
Should we mention `Ctrl` + scroll wheel here?

ericmehl replied:
Ah yes, of course. I also tweaked the wording just a bit - `Cursor up/down or <kbd>Ctrl</kbd> + scroll wheel to increment/decrement`. Having "up/down" right after "cursor" feels like a clearer explanation of the keyboard part.

ericmehl replied:
That's done in a6e6cbfd415643e829e42303b7e34b29a372cf70, where I squashed everything down.

ericmehl commented:
I fixed the wording of the tooltip and squashed everything down to one commit. Hopefully ready to merge!

murraystevenson commented:
Thanks Eric! Merging.
